### PR TITLE
Logistic regression (log-linear data) simulation experiments

### PR DESCRIPTION
**Pull request by jjc2718**
The goal in this PR is to simulate (X, y) data from a log-linear model (logistic regression) in which some of the features are correlated with the labels, then use L1-penalized logistic regression to see if we can find the features used to generate the labels.

We also did some experiments where we added duplicate features (exact copies of some of the randomly generated features) to the model, to see if that would bring performance down. As expected, it does.

Planning to generalize this code in the future to generating labels from features with varying degrees of collinearity, in order to benchmark e.g. network-regularized methods.

### CODE REVIEW

File: simdata/simulate_data_l1.ipynb [hunk 1]
```diff
@@ -0,0 +1,1061 @@
...
...9/fwDFHaRLly6Fq6srxGIxXFxcMGDAAADFw7fnzJmDRYsWwcbGRojHwMAA2traiIuLQ9u2bTFixAgkJSXBxcWlwovViujp6SEwMBArV65EYWEhiAgrV66Evr6+0GH0Lg0NDcyfPx+zZ8+GhoYGRCIRfvjhB7lj0a5dO/j5+WHq1KmQSqXQ0dFBcHBwuU89yys3ZZNufgx+fn5YuHAhXFxcIBKJsHLlSiFuZW3dyJEjkZSUBDc3N0gkEgwfPhyWlpYAoHRiO6B41N+xY8ewevVq6OjoYNOmTX/rydqHjh0ovkAfPXr0e8fEqlfv3r2xY8cO4eFN9+7dERUVJTcBJVB8/q1evRqzZ8+GpqYmunbtKqTZ2NgovY4szd7eHlFRURgyZAjq1KmDDh06CBOaz5w5E1OmTEHNmjVRq1YtdO3aVWhXHR0dsXbtWkgkEqX5hg0bhn79+mHIkCGoWbMmdHR0sHDhQmhpaWHz5s1YtmwZtm/fjqKiInz77bfCpIYl1+9BQUFynYHdu3fHggULkJWVhTp16mDatGnw8/NDv379oKen994dh0Dlv4feLX9l7V8Ja2treHp6wsPDAzKZDHp6eti6dWu5I6A6dOhQZrkB5bdNH9ratWvh6+uLLVu2QEtLCxs2bICamtp7X/8uWLAAxsbGctekJT777DOsX78eqamp0NPTw8qVK1UqdmXX7kDxT6y3aNHik3vIJ6LKvmvA2AcmlUoxePBgpTe9TLV5e3sLv7zy7uelf4nhYxszZgzmz59fqWF87P+fs2fPYu/evcJ7zqxsJ06cwJ07d7Bo0aLqDoVVUUJCAg4dOoTZs2fLfZ6SkgJXV1fcu3evmiKrWFJSEmbPno0DBw6UOQcGY5+y4OBgqKurK8zBw1Tf9evXkZSUpNBhUfqXCz9F3t7e6Nu3L3r06FHdoVQJvxLCqo26ujqWLl1a4asaTPXk5+fDyspKobPin3b+/Hl06dKFOysY+5tcXV2RmZlZ5uSaTLXFx8eXOaHxp2D9+vXw9/fnzgr2rzR27FjcvHkTr169qu5QWBW9fftWmLD43yIyMhIikeiT66wAeIQFY4wxxhhjjDHGVBCPsGCMMcYYY4wxxpjK+SQn3ZTJZMjNzVX4STXGGPs3ICJIJBLo6uqq1M/9cdvLGPs347aXMcb+eRW1vZ9kh0Vubi6io6OrOwzGGPuoDA
```
Comment (ben-heil) [inline simdata/simulate_data_l1.ipynb]: The l1 regularization term here probably isn't equivalent to the one used in the pytorch LR implementation. That being said, I don't think your point in including both algorithms here is to compare them, so it's not a big deal. Also it's surprisingly difficult to find exactly how the `C` parameter in sklearn.LogisticRegression is used past that it's an inverse regularization term

Comment (ben-heil) [inline simdata/simulate_data_l1.ipynb]: Nice! Weird that the sklearn logistic regression is consistently doing worse than random chance on the tall matrices though

File: simdata/simulate_loglinear.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    n_test = n - n_train
+
+    if verbose:
+        print('Train samples: {}, test samples: {}'.format(n_train, n_test))
+
+    train_ixs = np.zeros(n).astype('bool')
+    train_ixs[:n_train] = True
+    np.random.shuffle(train_ixs)
+
+    return train_ixs
```
Comment (ben-heil) [inline simdata/simulate_loglinear.py]: In the future you may find it easier to use sklearn's built in [train test split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) function instead of building your own

File: utilities/classify_pytorch.py [hunk 3]
```diff
@@ -419,11 +429,10 @@ def torch_tuning(self, X_subtrain, X_tune, y_subtrain, y_tune):
     y_pred_train, y_pred_test = preds
     y_pred_bn_train, y_pred_bn_test = preds_bn
 
-    def calculate_accuracy(y, y_pred):
-        return sum(1 for i in range(len(y)) if y[i] == y_pred[i]) / len(y)
-
-    torch_train_acc = calculate_accuracy(y_train, y_pred_bn_train)
-    torch_test_acc = calculate_accuracy(y_test, y_pred_bn_test)
+    print(y_train.shape)
+    print(y_pred_bn_train.shape)
```
Comment (ben-heil) [inline utilities/classify_pytorch.py]: These may be left over debug print statements

File: simdata/simulate_data_l1.ipynb (hunk 1 above)
Comment (jjc2718) [reply simdata/simulate_data_l1.ipynb]: That's true. I also think we're using elastic net in sklearn currently (and no L2 penalty in PyTorch), so it's not meant to be a perfect comparison. The performance of the two implementations is roughly the same, so I'm not too worried about the details matching perfectly.

File: simdata/simulate_data_l1.ipynb [hunk 4]
```diff
@@ -0,0 +1,1061 @@
...
+      "Torch coefficient Jaccard similarity: 0.231 (3/13)\n"
+     ]
+    }
+   ],
+   "source": [
+    "torch_coefs = torch_model.last_weights.flatten()\n",
+    "close_to_zero = (np.abs(torch_coefs) < (np.abs(torch_coefs).max() / 100))\n",
+    "torch_coefs[close_to_zero] = 0\n",
+    "sk_jaccard = compute_jaccard(np.where(~is_correlated)[0], np.where(clf.coef_.flatten() == 0)[0])\n",
+    "torch_jaccard = compute_jaccard(np.where(~is_correlated)[0], np.where(torch_coefs == 0)[0])\n",
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply simdata/simulate_data_l1.ipynb]: Exactly - I'll add a comment.

File: simdata/simulate_loglinear.py (hunk 2 above)
Comment (jjc2718) [reply simdata/simulate_loglinear.py]: Will add this to my refactoring issue (#8)
